### ISSUE TITLE
Different behaviour for regular and static pods upon deletion

### ISSUE THREAD

**Issue by harche** [kubernetes/kubernetes#114986]
### What happened?

When the regular pod is deleted it's status immediately transitions into `Terminating`, however when the static pod is deleted it stays in `Running` status until almost till the end just before it gets wiped out. 

### What did you expect to happen?

The reported status of the regular pod and static pod should be identical after deletion. 

### How can we reproduce it (as minimally and precisely as possible)?

Sample pod, 

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test 
spec:
  containers:
  - name: test 
    image: busybox:latest 
    command:
    - sleep 
    - inf 
  terminationGracePeriodSeconds: 30
  restartPolicy: Never
```

Scenario 1, 

Deploy it as a regular pod. e.g. `kubectl create -f test.yaml` and once it's running delete it using `kubectl delete -f test.yaml`. The prompt won't return until the pod is actually deleted, so from another terminal check the pod status,

```bash
$ kubectl get pods
No resources found in default namespace.
$ kubectl get pods
NAME   READY   STATUS    RESTARTS   AGE
test   1/1     Running   0          6s
$ kubectl get pods
NAME   READY   STATUS        RESTARTS   AGE
test   1/1     Terminating   0          12s
$ kubectl get pods
NAME   READY   STATUS        RESTARTS   AGE
test   1/1     Terminating   0          19s
$ kubectl get pods
NAME   READY   STATUS        RESTARTS   AGE
test   1/1     Terminating   0          23s
$ kubectl get pods
NAME   READY   STATUS        RESTARTS   AGE
test   1/1     Terminating   0          25s
$ kubectl get pods
NAME   READY   STATUS        RESTARTS   AGE
test   1/1     Terminating   0          29s
$ kubectl get pods
NAME   READY   STATUS        RESTARTS   AGE
test   1/1     Terminating   0          32s
$ kubectl get pods
NAME   READY   STATUS        RESTARTS   AGE
test   1/1     Terminating   0          39s
```

As you can see the pod immediately goes in `Terminating` state. 

Scenario 2,

Now deploy the same pod as a static pod by, `cp test.yaml /var/run/kubernetes/static-pods/` and after it starts running remove it using `rm /var/run/kubernetes/static-pods/test.yaml`. However in this case, the pod doesn't immediately go in `Terminating` state. Instead, it stays in `Running` until a last split second where it transitions into `Terminating` just before getting wiped out. 

```bash
$ kubectl get pods
No resources found in default namespace.
$ kubectl get pods
NAME             READY   STATUS    RESTARTS   AGE
test-127.0.0.1   1/1     Running   0          3s
$ kubectl get pods
NAME             READY   STATUS    RESTARTS   AGE
test-127.0.0.1   1/1     Running   0          22s
$ kubectl get pods
NAME             READY   STATUS    RESTARTS   AGE
test-127.0.0.1   1/1     Running   0          24s
$ kubectl get pods
NAME             READY   STATUS    RESTARTS   AGE
test-127.0.0.1   1/1     Running   0          25s
$ kubectl get pods
NAME             READY   STATUS    RESTARTS   AGE
test-127.0.0.1   1/1     Running   0          27s
$ kubectl get pods
NAME             READY   STATUS    RESTARTS   AGE
test-127.0.0.1   1/1     Running   0          29s
$ kubectl get pods -w
NAME             READY   STATUS    RESTARTS   AGE
test-127.0.0.1   1/1     Running   0          33s
$ kubectl get pods 
NAME             READY   STATUS    RESTARTS   AGE
test-127.0.0.1   1/1     Running   0          42s
$ kubectl get pods -w
NAME             READY   STATUS    RESTARTS   AGE
test-127.0.0.1   1/1     Running   0          49s
test-127.0.0.1   0/1     Error     0          50s
test-127.0.0.1   0/1     Terminating   0          51s
test-127.0.0.1   0/1     Terminating   0          51s
```
 

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by harche** [comment]
It could be kubelet issue or the issue with how kubectl reports the status. Not sure at this point.

/sig node
/sig cli

**Comment by SergeyKanzhelev** [comment]
keep it untriaged to understand if this is a cli or kubelet issue. If it is just a visualization issue on cli, priority would be different

**Comment by harche** [comment]
@Sajiyah-Salat Sure, thanks.

/assign @Sajiyah-Salat

**Comment by Sajiyah-Salat** [comment]
Can we connect somewhere to discuss about it.

**Comment by harche** [comment]
As a first step we need to decide if it's just issue with `kubectl` or really something isn't quite right within kubelet. I would start with tracing how `kubectl get pods` lists the pods and see how to proceed from there.

**Comment by Sajiyah-Salat** [comment]
Great. Let me know your social handle.

**Comment by harche** [comment]
We usually use github for collaboration, so I would prefer to stick with that.

**Comment by Sajiyah-Salat** [comment]
Great then. Any updates?

**Comment by SergeyKanzhelev** [comment]
@Sajiyah-Salat do you plan to work on this and try to repro it as @harche suggested in https://github.com/kubernetes/kubernetes/issues/114986#issuecomment-1381984600. If we can stick with updating this issue with relevant finding instead of moving conversation to Slack - it will be useful for everybody.

**Comment by Sajiyah-Salat** [comment]
I will try to get some findings

**Comment by SergeyKanzhelev** [comment]
/triage accepted

@Sajiyah-Salat myself and @harche are confused whether you plan to work on this. Please let us know.

**Comment by Sajiyah-Salat** [comment]
Sorry @SergeyKanzhelev for the late reply. I tried to reproduce but I have minikube on my system. and while running command ``cp test.yaml /var/run/kubernetes/static-pods/`` I got this error ``cp: cannot create regular file '/var/run/kubernetes/static-pods/': No such file or directory`` I tried in killercoda and got the same error.

**Comment by Sajiyah-Salat** [comment]
In this issue the author has **deleted**  the regular pod with this command
``k delete -f test.yaml``
while in static pod he had **remove** the static pod with this command
``rm test.yaml``
Dont you guys think that both command works differently?

**Comment by harche** [comment]
@Sajiyah-Salat you can read more on static pods here, https://kubernetes.io/docs/tasks/configure-pod-container/static-pod/

BTW, would it be okay with you if I assign this issue to me?

**Comment by harche** [comment]
I just verified that upcoming pod manager refactor, https://github.com/kubernetes/kubernetes/pull/115342, has no effect on this issue.

**Comment by harche** [comment]
> I just verified that upcoming pod manager refactor, #115342, has no effect on this issue.

Although it's not a complete code and still work in progress.

**Comment by smarterclayton** [comment]
Copying from #114994 (I noticed this while looking at e2e tests for #113145 that @bobbypage was adding):

We discussed adding conditions to help api server clients understand when a pod had released its resources (right now there is no unambiguous signal that callers can safely use other than the transition to succeded or failed state, which is blocked by a bug on pending pods not going into a terminal phase that @bobbypage was gonna help the Job termination folks do...). We might want to consider one for indicating when a mirror pod is going away, OR we could just use graceful deletion with a finalizer tied to the kubelet and have the kubelet do the removal once the static pod is gone.

Static mirror pods need to behave like regular pods during deletion, update, and creation IMO. I think that includes readiness, pending termination and others.

One scenario we can't fix is if you reuse your static pod UID there's no way to represent "completely changing" in the API except via the spec update (which I think gets blocked?). Now I have to go figure out what that does...

**Comment by rajeeshckr** [comment]
We have self managed kubernetes cluster. We are running the API-server as a mirror pod. During shutdown of node, we wanted the API-server mirror pod to shutdown gracefully. Since [graceful node termination](https://kubernetes.io/blog/2021/04/21/graceful-node-shutdown-beta/#how-do-i-use-it) handles only regular pods, we can't use this approach.
We are using the systemd hooks to shutdown the API mirror pod gracefully once kubelet is shutting down on this node to get around this issue.
Will this scenario can also be considered while working on this? Or we need a new issue?

**Comment by harche** [comment]
fixed by https://github.com/kubernetes/kubernetes/pull/116995 

/close
